src/amf3/encode.rs

```diff
@@ -15,6 +15,10 @@ impl<W> Encoder<W> {
     pub fn into_inner(self) -> W {
         self.inner
     }
+    /// Returns a reference to the underlying writer.
+    pub fn inner(&mut self) -> &mut W {
```

Comment:
I prefer `inner_mut` as the method name because it returns a mutable reference.
How about renaming this method and adding another method, named `inner`, that returns a immutable reference?

Reply:
Alright, I'll do that.
I can also submit a separate pull request to change  it on Decoder.

Reply:
Hmmm....
How about having `ReadAmfExt` and `WriteAmfExt` traits instead of `Encoder` and `Decoder`?
Or just have `inner` be public? Or would you rather have access to the inner reader/writer be more controlled?

Reply:
Actually, Decoder needs additional state, so an extension trait won't do it for that.
How about having `inner` be public though?

Reply:
> just have inner be public? 

It sounds better choice than others.
Through `inner_mut` method, users could violate the decoding/encoding process, but it's responsibility to them to use the method correctly (it's the same convention, e.g., as [BugReader.get_mut](https://doc.rust-lang.org/std/io/struct.BufReader.html#method.get_mut)).